File: client/concrete_client.cc
//
// Created by dylan on 10/08/2020.
//

#include "concrete_client.h"
#include "client.h"
#include "util/protobuf_util.h"
#include "util/platform/reader.h"
#include "util/mapping_reader.h"

#include <memory>
#include <sstream>

namespace tetris {

ConcreteClient::ConcreteClient(const std::string &server_socket_path,
        const std::string &platform_desc_path, const std::string &mapping_path)
        : Client(), _push_message_listener(get_push_listener_socket_path(), this),
          _managed(false), _communication_mutex()
{
    _logger = debug::Logger::get();
    try {
        /* Read the platform file */
        YamlPlatformReader platform_reader;
        _platform = std::move(platform_reader.ReadFromFile(platform_desc_path));

        /* Read the mappings */
        YamlMappingReader mapping_reader;
        _mappings = std::move(mapping_reader.read_mappings(*_platform, mapping_path));
        _logger->debug(" -> Loaded %d mappings for this client\n", _mappings.size());

        _tetris_server_connection.connect(server_socket_path);
        _managed = register_client();

        if (_managed) {
            /* Send over the mappings to the server, so that we can get scheduled */
            ClientMessage msg;
            msg.set_type(ClientMessage::OPERATING_POINTS);
            auto ops_info = msg.mutable_ops_info();

            for (const auto& m : _mappings) {
                auto op_data = ops_info->add_operating_points();

                op_data->set_identifier(m.name);
                for (const auto& [cn, cv] : m.characteristics_map) {
                    auto c = op_data->add_characteristics();
                    c->set_name(cn);
                    c->set_value(cv);
                }

                for (const auto& c : m.cpus) {
                    op_data->add_cpu_ids(c);
                }
            }

            ServerResponse response{};
            _communication_mutex.lock();
            protobuf_util::Send(_tetris_server_connection.locked(), msg);
            protobuf_util::Receive(_tetris_server_connection.locked(), response);
            _communication_mutex.unlock();

            if (response.type() != ServerResponse::ACKNOWLEDGE) {
                _logger->warning("The server failed to parse our mappings!\n");
            }
        }
    } catch (std::exception &e) {
        _logger->info("No TETRiS server, TETRiS is unused.\n");
        _managed = false;
    }
}

void ConcreteClient::bind(Feature *feature)
{
    // If the client is not connected to the server, do not bind the feature.
    if (!_managed)
        return;
    // Bind the client to the feature.
    feature->accept(this);
    // If it needs a handshake, perform it.
    if (feature->need_handshake()) {
        auto feature_id = feature->handshake();
        // Add the feature to the subscriber lists of the push listener.
        _push_message_listener.add_subscriber(feature_id, feature);
    }
}

void ConcreteClient::bind(MappingFeature *feature)
{
    if (!_managed)
        return;

    feature->accept(this);

    if (feature->need_handshake()) {
        auto feature_id = feature->handshake();

        _push_message_listener.add_subscriber(feature_id, feature);
    }

    _mapping_features.push_back(feature);

    if (_active_mapping)
        feature->mapping_update(*_active_mapping, {});
}

ServerResponse ConcreteClient::send(const ClientMessage &message)
{
    ServerResponse response{};
    _communication_mutex.lock();
    protobuf_util::Send(_tetris_server_connection.locked(), message);
    protobuf_util::Receive(_tetris_server_connection.locked(), response);
    _communication_mutex.unlock();
    return response;
}

std::string ConcreteClient::get_push_listener_socket_path()
{
    std::stringstream string_stream{};
    string_stream << "/tmp/tetris_push_listener_" << getpid();
    return string_stream.str();
}

ClientResponse ConcreteClient::handle(const ServerMessage &msg)
{
    ClientResponse response{};
    response.set_type(tetris::ClientResponse::ERROR);

    if (msg.has_activated_op_info()) {
        /* Call all mapping_features with the new mapping, so that they can adapt */
        auto active_op = msg.activated_op_info();

        auto map_id = active_op.identifier();
        _logger->info(" * Got mapping update from server: %s\n", map_id.c_str());

        std::map<int, int> conv_map;
        for (int i = 0; i < active_op.cpu_convs_size(); ++i) {
            auto conv = active_op.cpu_convs(i);
            conv_map.emplace(conv.cpu_id_from(), conv.cpu_id_to());
        }

        auto it = std::find_if(_mappings.begin(), _mappings.end(), [&map_id](const Mapping& m) { return m.name == map_id; });
        if (it != _mappings.end()) {
            _active_mapping = std::make_unique<Mapping>(*it, conv_map);
            _logger->debug(" -> Active mapping %s\n", _active_mapping->name.c_str());

            /* Tell the features to react to the new mapping */
            for (const auto& feature : _mapping_features) {
                feature->mapping_update(*_active_mapping, conv_map);
            }

            response.set_type(ClientResponse::ACKNOWLEDGE);
        }
    }

    return response;
}

bool ConcreteClient::register_client()
{
    RegistrationRequest request{};

    /* Send the new-client message to the server. */
    request.set_pid(getpid());
    char exec[512];
    memset(exec, 0, sizeof(exec));
    readlink("/proc/self/exe", exec, sizeof(exec));
    request.set_exec(exec);

    // Send the command.
    try {
        RegistrationResponse response{};
        _communication_mutex.lock();
        protobuf_util::Send(_tetris_server_connection.locked(), request);
        protobuf_util::Receive(_tetris_server_connection.locked(), response);
        _communication_mutex.unlock();

        // Process the TETRiS server response.
        _logger->info("TETRIS-ID: %d\n", response.id());
        return true;
    } catch (std::exception &e) {
        return false;
    }
}

}

File: client_py/concreteClient.py
import logging
import os
import socket
import threading

from client_py.client import Client
from client_py.mappingFeature import MappingFeature
from client_py.push_message_listener import PushMessageListener
from client_py.utils.protobufUtil import ProtobufUtil
from client_py.utils.yamlMappingReader import YamlMappingReader
# from mappingFeature import MappingFeature
from proto.tetris_pb2 import ClientMessage, ServerResponse, RegistrationRequest, RegistrationResponse, ClientResponse, \
    ServerMessage


class ConcreteClient(Client):

    def __init__(self, server_socket_path, platform_desc_path, mapping_path):
        super().__init__()
        self._managed = False
        self._logger = logging.getLogger('ConcreteClient')
        self._push_message_listener = PushMessageListener(self.__get_push_listener_socket_path(), self)

        try:
            # Read the platform file
            # self._platform = YamlPlatformReader.read_from_file(platform_desc_path)

            # Read the mappings
            self._mappings = YamlMappingReader.read_mappings(file_path=mapping_path)
            self._active_mapping = None

            self._mapping_features = []
            # self._logger.debug(f" -> Loaded {len(self._mappings)} mappings for this client")

            # todo: check whether there is a lock even needed in socket-lib-usage
            self._communication_mutex = threading.Lock()

            self._tetris_server_connection = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
            self._tetris_server_connection.connect(server_socket_path)
            self._managed = self.register_client()

            # print(self._managed)
            if self._managed:
                # Send available mappings to the server
                msg = ClientMessage()
                msg.type = ClientMessage.OPERATING_POINTS
                self.__add_mappings_to_client_message(self._mappings, msg)

                ProtobufUtil.send(self._tetris_server_connection, msg)
                response = ServerResponse()
                ProtobufUtil.receive(self._tetris_server_connection, response)

                if response.type != ServerResponse.ACKNOWLEDGE:
                    self._logger.warning("The server failed to parse our mappings!")
                # self._logger.info("Server connection established. Operating points were sent successfully")
        except Exception as e:
            print("exception thrown: ", e)
            self._logger.info("No TETRiS server, TETRiS is unused.")
            self._managed = False

    @staticmethod
    def __add_mappings_to_client_message(mappings, msg):
        for mapping in mappings:
            op = msg.ops_info.operating_points.add()
            op.identifier = mapping.name
            op.cpu_ids.extend(mapping.cpu_ids)
            for name, value in mapping.characteristics.items():
                characteristic = op.characteristics.add()
                characteristic.name = name
                characteristic.value = value

    @staticmethod
    def __get_push_listener_socket_path():
        return f"/tmp/tetris_push_listener_{os.getpid()}"

    def bind(self, feature):
        # to implement if needed
        # but not in use atm.
        raise NotImplementedError("Implement method if needed!")

    def send(self, msg: ClientMessage) -> ServerResponse:
        # to implement if needed
        # but not in use atm.
        raise NotImplementedError("Implement method if needed!")

    def bind_mapping_feature(self, feature: MappingFeature):
        if not self._managed:
            return

        feature.accept(self)

        if feature.need_handshake():
            feature_id = feature.handshake()
            self._push_message_listener.add_subscriber(feature_id, feature)

        self._mapping_features.append(feature)

        if self._active_mapping:
            # todo: implement
            raise NotImplementedError("implement mapping update with conv-map")
            feature.mapping_update(self._active_mapping, {})

    def handle(self, msg: ServerMessage) -> ClientResponse:
        response = ClientResponse()
        response.type = ClientResponse.Type.ERROR

        if hasattr(msg, 'activated_op_info'):
            active_op = msg.activated_op_info
            map_id = active_op.identifier
            print(f" * Got mapping update from server: {map_id}")

            conv_map = {conv.cpu_id_from: conv.cpu_id_to for conv in active_op.cpu_convs}

            # Search for the mapping with the given map_id
            mapping = next((m for m in self._mappings if m.name == map_id), None)

            if mapping:
                self._active_mapping = mapping
                print(f" -> Active mapping {self._active_mapping.name}")

                # Tell the features to react to the new mapping
                for feature in self._mapping_features:
                    feature.mapping_update(self._active_mapping, conv_map)

                response.type = ClientResponse.Type.ACKNOWLEDGE

        return response

    def register_client(self):
        request = RegistrationRequest()

        request.pid = os.getpid()
        request.exec = os.path.realpath(__file__)

        try:
            ProtobufUtil.send(self._tetris_server_connection, request)
            response = RegistrationResponse()
            ProtobufUtil.receive(self._tetris_server_connection, response)
            print(f"TETRIS-ID: {response.id}")
            return True
        except Exception as e:
            print("error", e)
            return False

    def close(self):
        self._tetris_server_connection.close()
